experimental/sev_guest/src/msr.rs

```diff
@@ -462,3 +462,174 @@ pub fn read_msr() -> u64 {
+        assert_eq!(expected, request.into());
+    }
+
+    #[test]
+    fn test_cpuid_response() {
+        let value = 7531u32;
+        let register = CpuidRegister::Ebx;
+        let invalid = 5u64 | ((value as u64) << 32) | ((register as u64) << 25);
+        let valid = 5u64 | ((value as u64) << 32) | ((register as u64) << 30);
+
+        let error: Result<CpuidResponse> = invalid.try_into();
+        assert!(error.is_err());
+
+        let correct: CpuidResponse = valid.try_into().unwrap();
+        assert_eq!(correct.value, value);
+        assert_eq!(correct.register, register);
+    }
+
+    #[test]
+    fn test_preferred_ghcb_gpa_request() {
+        let request = PreferredGhcbGpaRequest;
+        assert_eq!(16u64, request.into());
+    }
+
+    #[test]
+    fn test_preferred_ghcb_gpa_response() {
+        let page = 97531u64 << 12;
+        let invalid = 0xFu64 | page;
+        let valid = 0x11u64 | page;
+
+        let error: Result<PreferredGhcbGpaResponse> = invalid.try_into();
+        assert!(error.is_err());
+
+        let correct: PreferredGhcbGpaResponse = valid.try_into().unwrap();
+        assert_eq!(correct.ghcb_gpa, page as usize);
+    }
+
+    #[test]
+    fn test_register_ghcb_gpa_request() {
+        // Page must be 4KiB aligned.
+        let invalid_page = 97531usize << 11;
+        assert!(RegisterGhcbGpaRequest::new(invalid_page).is_err());
+
+        let valid_page = 97531usize << 12;
+        let msr_value: u64 = RegisterGhcbGpaRequest::new(valid_page).unwrap().into();
+        assert_eq!((valid_page as u64) | 0x12, msr_value);
+    }
+
+    #[test]
+    fn test_register_ghcb_gpa_response() {
+        let page = 222221u64 << 12;
+        let invalid = 0x12u64 | page;
+        let valid = 0x13u64 | page;
+
+        let error: Result<RegisterGhcbGpaResponse> = invalid.try_into();
+        assert!(error.is_err());
+
+        let correct: RegisterGhcbGpaResponse = valid.try_into().unwrap();
+        assert_eq!(correct.ghcb_gpa, page);
+    }
+
+    #[test]
+    fn test_snp_page_state_change_request() {
+        let assignment = PageAssignment::Shared;
+        // Page must be 4KiB aligned.
+        let invalid_page = 1111111usize << 11;
+        assert!(SnpPageStateChangeRequest::new(invalid_page, assignment).is_err());
+        // Page address can only use 52 bits.
+        let invalid_page = 1usize << 52;
+        assert!(SnpPageStateChangeRequest::new(invalid_page, assignment).is_err());
+
+        let valid_page = 1111111usize << 12;
+        let msr_value: u64 = SnpPageStateChangeRequest::new(valid_page, assignment)
+            .unwrap()
+            .into();
+        assert_eq!(
+            (valid_page as u64) | ((assignment as u64) << 52) | 0x14,
+            msr_value
+        );
+    }
+
+    #[test]
+    fn test_snp_page_state_change_response() {
+        let error_code = 1u64;
+        let invalid = 0x15u64 | (error_code << 31);
+        let invalid2 = 0x16u64 | (error_code << 32);
```

Comment:
Is it possible to give it a better name?

Reply:
Done.